feat: Add NATS Core + JetStream implementations

AlexCuse opened the pull request:
Introduces outline for NATS core client as well

<!-- Expected Commit Message Description (imported automatically by GitHub) -->
<!-- Must conform to [conventional commits guidelines](https://github.com/edgexfoundry/go-mod-messaging/blob/main/.github/Contributing.md) -->
<!-- Expected Commit message must contain Closes/Fixes #IssueNumber statement when there is a related issue -->

<!-- Add additional detailed description of need for change if no related issue -->

If your build fails due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/go-mod-messaging/blob/main/.github/Contributing.md

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [ ] I am not introducing a new dependency (add notes below if you are)
       -  adding https://github.com/nats-io/nats.go (apache2) and https://github.com/hashicorp/go-multierror (mpl-2)
- [x] I have added unit tests for the new feature or bug fix (if not, why?)
- [x] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
       hope to look at adding coverage via TAF once in main
- [x] I have opened a PR for the related docs change (if not, why?)
       https://github.com/edgexfoundry/edgex-docs/pull/805

## Testing Instructions
<!-- How can the reviewers test your change? -->
There is an example included that should be easy to follow.

## New Dependency Instructions (If applicable)
<!-- Please follow [vetting instructions](https://wiki.edgexfoundry.org/display/FA/Vetting+Process+for+3rd+Party+Dependencies) and place results here -->

AlexCuse commented:
A couple notes:

- I have not tested any of the auth modes yet but unauthenticated is working
- I have not really tested the NATS core client either but noticed the actual nats.go client design seemed to allow the approach.  I think there is probably a set of default settings that will work here.
- I am not sure the convention of appending ID to the end of the subject is actually necessary, if we wanted to eliminate could use the NATS header to transmit CorrelationID as well.

AlexCuse commented:
The nats core client may actually be in better shape than I thought, seemed to be able to get some data moving through it.  But my focus has been on jetstream.

AlexCuse commented:
I removed the `correlation id in subject` format, same can be achieved using edgex topic replacements.

Outstanding work I see includes:

- [ ] testing authentication
- [ ] option builders

I would like to figure out how to add to any existing tests in edgex-taf around the message bus as well.

lenny-intel commented:
@AlexCuse you are adding `github.com/nats-io/nats.go` which is a new dependency and need to go thru the Vetting process. Please complete the PR template with this information (which you have left blank).

lenny-intel commented:
@JamesKButcher, @jpwhitemn , Previous CoreData size w/o this addition is `27,554,912` bytes, Core Data size with this addition is `29,141,576` bytes. That is a delta of an additional  `1,586,664` bytes or `1.5` MB.

AlexCuse commented:
I updated PR for the direct dependencies Lenny.  

That size is considerable - I had not thought of this.  Are you looking at the built docker image size or the binary size that gets placed on it?  I think most of the baggage in the NATS client comes from the connection encoders that we don't necessarily use, wonder if I can make a case over there for deprecating or moving to a different package.  I will try to get the same number for a custom package built without the encoders to make it.

AlexCuse commented:
@lenny-intel if you get a chance can you try your test build for size with:

`replace github.com/nats-io/nats.go => github.com/AlexCuse/nats.go v1.14.1-0.20220410151949-661249555d4e`

This is a quick and dirty attempt to just delete the encoders, I've confirmed it still works with my test harness and it removes both of the protobuf dependencies that I **suspect** are the source of the bloat.  If this works I can start a dialog with the nats team about moving encoders to a separate package from the main client.  I believe I saw that the encoders are not supported for JetStream at some point so there might already be some momentum behind the move.

AlexCuse commented:
Noticed a comment on this issue today about deprecating the encoders and moving to a new package: https://github.com/nats-io/nats.go/pull/954#issuecomment-1095230198

AlexCuse commented:
Not made much progress yet in shrinking the NATS client using https://github.com/jondot/goweight 

Out of curiosity do we know what is the cost of the zeromq implementation is (including libzmq etc dependencies from outside)

Maybe this could be a fair trade off down the line.

jpwhitemn commented:
@AlexCuse - Alex - the TSC and the community are exploring scope for the Levski release (fall 2022).  We are looking into this PR and ideas of whether to officially support NATS in EdgeX.  Would you be willing to make a presentation to either the TSC or Core working group in the near future (over the next several weeks) to cover:
a. details of your use case and why you believe NATS is important and should be used over Redis Pub/Sub or MQTT for the message bus implementation
b. What do you see as the pros/cons of NATs over Redis P/S or MQTT
c. Why JetStream implementation

If you are willing to present, let me know what timezone you are in and what weeks you might be available.  Send to jim@iotechsys.com
thx

AlexCuse commented:
Something I will do here when I rebase is remove the auth methods beyond "none" and "username/password" to keep the initial PR simple.  Then we can add TLS and others as proper dedicated efforts and a chance to iterate on client bootstrapping code @lenny-intel.

AlexCuse commented:
@lenny-intel I removed all the auth options other than username/password.  This should be good to merge and refine through the release cycle.

I will create a ticket for the docs change and assign myself.

lenny-intel commented on internal/pkg/nats/client.go:
```diff
@@ -0,0 +1,180 @@
+			if err != nil {
+				messageErrors <- err
+			} else {
+				tc.Messages <- env
```
Need to set ReceivedTopic on MessageEnvelope. It needs to be the actual topic that was publish, not the subscribe topic with wild cards. How do we get that from NATS?

AlexCuse replied:
The unmarshalers handle getting this directly from the NATS message itself.

lenny-intel reviewed (changes requested):
First pass review. Still need to finish reviewing client tests and the jetstream code

AlexCuse commented on _example/README.md:
```diff
@@ -0,0 +1,9 @@
+# example
```
removed, I just had it here to have an easy way to run the changed code.

AlexCuse commented on internal/pkg/nats/client_test.go:
```diff
@@ -0,0 +1,333 @@
+
+		nm := nats.NewMsg("topic.id")
+
+		connection.On("PublishMsg", nm).Return(fmt.Errorf("err"))
```
This test is to validate behavior when the underlying nats client returns an error, the marshaler is setup to return `nm`

AlexCuse commented on Makefile:
```diff
@@ -7,7 +7,7 @@ tidy:
 
 unittest:
-	$(GO) test -race ./... -coverprofile=coverage.out ./...
+	$(GO) test -tags=$(TAGS) -race ./... -coverprofile=coverage.out ./...
```
set it by default

AlexCuse commented on Makefile:
```diff
@@ -18,5 +18,10 @@ test: unittest lint
 	gofmt -l $$(find . -type f -name '*.go'| grep -v "/vendor/")
 	[ "`gofmt -l $$(find . -type f -name '*.go'| grep -v "/vendor/")`" = "" ]
 
+BENCHCNT := 5
+
+bench:
+	go test -bench=. -count $(BENCHCNT) -run=^# ./...
+
```
Added benchmarks for the marshalers should I remove?

lenny-intel replied:
Nope, but a comment that that is what it is useful for will help.

AlexCuse commented on internal/pkg/nats/README.md:
```diff
@@ -0,0 +1,45 @@
+messagebus trigger via `go-mod-messaging`.  A [custom trigger](https://docs.edgexfoundry.org/2.0/microservices/application/Triggers/#custom-triggers) may be an
+option if finer grained control of the NATS configuration is needed.
+
+All option are set via `MessageBusConfig.Optional` and keys must be all lowercase.
```
good idea I will look at how some of this could be reused along with the constants.

AlexCuse commented on internal/pkg/nats/client.go:
```diff
@@ -0,0 +1,180 @@
+// NewClient initializes creates a new client using NATS core
+func NewClient(cfg types.MessageBusConfig) (*Client, error) {
+	return NewClientWithConnector(cfg, func(broker types.HostInfo, option ...nats.Option) (Connection, error) {
+		url := nats.DefaultURL
+		if !broker.IsHostInfoEmpty() {
+			url = broker.GetHostURL()
+		}
+
+		nc, err := nats.Connect(url, option...)
+
+		if err != nil {
+			return nil, err
+		}
+
+		return nc, nil
```
This is passed in to allow the same client logic to use a jetstream or core NATS connection - can look at other options.

lenny-intel replied:
Ok, see that it is used by both Core and JetStream...

AlexCuse commented on internal/pkg/nats/client.go:
```diff
@@ -0,0 +1,180 @@
+		return fmt.Errorf("connection function not specified")
+	}
+
+	subConnectOpt, err := buildSubscriberConnectOpt(c.config.SubscribeHost, c.config.Optional)
```
done

AlexCuse commented on internal/pkg/nats/connection.go:
```diff
@@ -0,0 +1,37 @@
+
+//go:build include_nats_messaging
+
+package nats
```
done

AlexCuse commented:
I think I took care of everything but the doc stuff and figuring out how to leverage the mqtt optionbuilder stuff - hopefully knock that out in the next week or so.

lenny-intel commented on internal/pkg/nats/options.go:
```diff
@@ -0,0 +1,151 @@
+	return ClientOptions{
+		Username: "",
+		Password: "",
+		// Client ID is required or else can cause unexpected errors. This was observed with Eclipse's Mosquito MQTT server.
```
Update this comment for NATS

lenny-intel reviewed (changes requested):
Here is my next pass. Still have more to review.

lenny-intel commented on internal/pkg/nats/client.go:
```diff
@@ -0,0 +1,180 @@
+
+// NewClientWithConnector creates a new client that uses the specified function to establish pub/sub connections
+func NewClientWithConnector(cfg types.MessageBusConfig, connector ConnectNats) (*Client, error) {
+	if connector == nil {
+		return nil, fmt.Errorf("connector is required")
+	}
```
NM

lenny-intel reviewed (changes requested):
First pass review of all files complete.

AlexCuse commented on internal/pkg/nats/jetstream/options.go:
```diff
@@ -0,0 +1,48 @@
+	durable = "durable"
+)
+
+func buildJSOpt(_ types.HostInfo, _ map[string]string) ([]nats.JSOpt, error) {
```
deleting these methods and just attaching placeholders.

AlexCuse commented on internal/pkg/nats/interfaces/mocks/Connection.go:
```diff
@@ -0,0 +1,62 @@
+// Code (mostly) generated by mockery v2.9.4. DO NOT EDIT.
```
the trick I need to work out here is that the interface uses NATS types that mockery doesn't import.

Pure generated code looks like this (note the invalid path to nats-io/nats.go):

```
// Code generated by mockery v2.13.1. DO NOT EDIT.

package mocks

import (
	nats "github.com/nats-io/"
	mock "github.com/stretchr/testify/mock"
)
```

lenny-intel replied:
OK, interesting that is doesn't find the NATS types. :-( And I see you are using the latest mockery.

AlexCuse commented on internal/pkg/nats/options.go:
```diff
@@ -0,0 +1,151 @@
+func (cc ClientConfig) connectOpt(ct connectType) []nats.Option {
+	connectTimeout := time.Second
+
+	if cc.ConnectTimeout != 0 {
```
Yea might be able to do millisecond even it just needs to be positive to attempt I think.  Since ints=seconds for the MQTT stuff thats sort of the line of thought that got me to a second.

lenny-intel commented on internal/pkg/nats/interfaces/mocks/Connection.go:
```diff
@@ -0,0 +1,81 @@
+package mocks
+
+import (
+	nats "github.com/nats-io/nats.go"
```
Never seen an import with the `.go` suffix

lenny-intel commented on internal/pkg/nats/options.go:
```diff
@@ -0,0 +1,138 @@
+		ConnectTimeout:          5, // 5 seconds
+		RetryOnFailedConnect:    false,
+		Durable:                 false,
+		AutoProvision:           false, // AutoProvision JetStream streams - should maybe be true?
```
Defaults should be in sync with the NATS defaults if not specified.

AlexCuse replied:
Default is false - problem there is you then can't publish to a topic until something is listening to it (this is an issue with core NATS also since there is no persistence layer also, there is just no way to auto-provision around it like jetstream provides).

lenny-intel replied:
Interesting, so we will get an error on publish if no subscriber?

AlexCuse commented on internal/pkg/nats/client_test.go:
```diff
@@ -0,0 +1,333 @@
+
+	t.Run("subscribe connector errors", func(t *testing.T) {
+		sut.connect = func(info types.HostInfo, option ...nats.Option) (Connection, error) {
+			return nil, fmt.Errorf("err")
```
I saw the note on deprecating the dual publish/subscribe hosts in edgex 3 going to make this same either/or proposition as with MQTT

lenny-intel commented on internal/pkg/nats/options.go:
```diff
@@ -0,0 +1,135 @@
+	Username string
+	Password string
+	ClientId string
+	Format   string
+	// Connection information
+	// TODO: NATS core is QOS 0, Jetstream QOS 1 (or 2 with sync ack + msgid)
+	// Qos                  int
```
Remove. Just do it the NATS way only

lenny-intel commented on internal/pkg/nats/jetstream/connection.go:
```diff
@@ -0,0 +1,80 @@
+	js               nats.JetStreamContext
+	durableSubscribe bool
+	subOpts          []nats.SubOpt
+	pubOpts          []nats.PubOpt
```
I see where this is used, but not where it is set???

AlexCuse replied:
Set when initializing the jetstream connection, I tried to clean this up a little bit.

AlexCuse commented on internal/pkg/nats/client.go:
```diff
@@ -0,0 +1,151 @@
+		return nil, err
+	}
+
+	switch strings.ToLower(cc.Format) {
+	case "nats":
+		m = &natsMarshaller{}
+	case "json":
+		m = &jsonMarshaller{}
+	case "cbor":
+		m = &cborMarshaller{}
+	}
```
One argument for the other marshalers could be if someone was using the MQTT support (https://docs.nats.io/running-a-nats-service/configuration/mqtt ) to connect legacy services.  Since the other services would need to be able to speak in JSON/CBOR envelopes over MQTT it could make sense to use the other format to enable NATS throughout as an alternative to some kind of adapter service that handles the encoding change.  I wonder if nats should be default though 🤔

lenny-intel replied:
Since this is for our internal EdgeX MessageBus, the `connect legacy services` scenario is out of scope. All Edgex Services (that use the MessageBus) must be configured to use the same MessageBus type (redis, mqtt, zero, or nats). 

Also, it is only the payload this is either JSON or CBOR encode by the client and passed in via the MessageEnvelope. The MessageEnvelope JSON encode in the other implementations, which results in the payload being double encoded. 

Lets keep is simple and take advantage of NATs capability to avoid the double encoding of the payload like we do with the other implementations.

lenny-intel replied:
Ok, I now see the need to have the two formats 

- "nats" for when all service have messagebus config set to `nats` 
- "edgex" for when other services have messagebus config set to `mqtt` and the broker is the NATS server in MQTT mode.
   - Note that this format is setting the NATs message to be the JSON encoded MessageEnvelope just like the other implementation do.

AlexCuse commented on internal/pkg/nats/interfaces/connection.go:
```diff
@@ -0,0 +1,35 @@
+
+import "github.com/nats-io/nats.go"
+
+type Msg = nats.Msg
+type MsgHandler = nats.MsgHandler
+type Subscription = nats.Subscription
```
this was fighting with mockery I don't think it helped, will remove.

AlexCuse commented:
Just rebased and pushed @lenny-intel - I think I addressed everything in the last round of comments.  Going to start looking at docs I think and plan to spend some time improving test coverage as I work through #147.  Have some other nice to have auth options I'd like to get to before fall also.  

Let me know if you think theres anything else critical to get into this first push.

lenny-intel commented on go.mod:
```diff
@@ -2,18 +2,28 @@ module github.com/edgexfoundry/go-mod-messaging/v2
 
 require (
 	github.com/eclipse/paho.mqtt.golang v1.4.1
+	github.com/fxamacker/cbor/v2 v2.4.0
```
Remove once removed CBOR encoding

lenny-intel commented on internal/pkg/nats/options.go:
```diff
@@ -0,0 +1,132 @@
+	Username             string
+	Password             string
+	ClientId             string
+	Format               string
```
Remove once only using NATS way to publish with use of header

AlexCuse commented on internal/pkg/nats/options.go:
```diff
@@ -0,0 +1,132 @@
+}
+
+func (cc ClientConfig) ConnectOpt() []nats.Option {
+	connectTimeout := time.Second
```
I'm taking this change with the others for now but going to look into the default option as well, theres a lot of benefits to aligning this better to the MQTT implementation I think.  If I remeber right you need to have it set to something with NATS.go to avoid it being zero but could be wrong or an issue thats since been fixed (I typically use 1s for local and 5s for NGS connections)

AlexCuse commented on internal/pkg/nats/jetstream/client.go:
```diff
@@ -0,0 +1,55 @@
+	// TODO: configure
+	durSub := true
+
+	return &connection{cc, conn, js, durSub, subOpts, pubOpts}, nil
```
isn't this inside the factory function?  Does it need to be public?

lenny-intel replied:
Sorry, missed that... ;-)

AlexCuse commented on internal/pkg/nats/client_test.go:
```diff
@@ -0,0 +1,301 @@
+
+	t.Run("happy", func(t *testing.T) {
+		sut.connect = func(ClientConfig) (interfaces.Connection, error) {
+			return &mocks2.Connection{}, nil
```
this connect function is used inside the public `Connect` - it is expected to return a live connection so nothing else needs to be called.

lenny-intel reviewed (approved):
LGTM
